modified to support both network based and serial based Onkyo receivers

basriram opened the pull request.

watou commented on bundles/binding/org.openhab.binding.onkyo/META-INF/MANIFEST.MF:
```diff
@@ -10,7 +10,9 @@ Bundle-Activator: org.openhab.binding.onkyo.internal.OnkyoActivator
  omation Bus (openHAB)
-Import-Package: javax.xml.bind,
+Import-Package: gnu.io,
+ org.apache.commons.io;version="2.0.1",
```
Please remove `;version="2.0.1"`

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpInterface.java:
```diff
@@ -0,0 +1,83 @@
+import org.openhab.binding.onkyo.internal.OnkyoStatusUpdateEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
The only `import` needed in this file is 
```java
import org.openhab.binding.onkyo.internal.OnkyoEventListener;
```
correct?  Also, please use the [latest IDE configuration](https://github.com/openhab/openhab-distro/blob/master/docs/sources/development/ide.md) to ensure proper code formatting and automatic cleanup of things like this. :smile:

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpInterface.java:
```diff
@@ -0,0 +1,83 @@
+ * @author Tom Gutwin P.Eng
+ * @author Thomas.Eichstaedt-Engelen (Refactoring)
+ * @author Pauli Anttila (Simplified, rewritten and added status update listener functionality)
+ * @author Sriram Balakrishnan (Added serial port support)
```
Since this is a new file, not as many authors as this?

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,261 @@
+ * Onkyo Connector for serial port communication.
+ * 
+ * @author Sriram Balakrishnan
+ * @since 1.8.0
```
1.9.0

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,261 @@
+		}
+	}
+	
+	private String sendMmsg(String data, int timeout) throws IOException {
```
Is `sendMmsg` the intended spelling, or should it be `sendMsg` or `sendMessage`?

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EispEventNotifier.java:
```diff
@@ -0,0 +1,100 @@
+package org.openhab.binding.onkyo.internal.eiscp;
```
Please add standard copyright header above this line:
```java
/**
 * Copyright (c) 2010-2016, openHAB.org and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
```

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EispEventNotifier.java:
```diff
@@ -0,0 +1,100 @@
+import org.openhab.binding.onkyo.internal.OnkyoEventListener;
+import org.openhab.binding.onkyo.internal.OnkyoStatusUpdateEvent;
+
+public class EispEventNotifier {
```
Please add JavaDoc comment block above this line, describing this class and your `@author` tag.

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,261 @@
+
+	
+			String response = sendMmsg(data, timeout);
+			if (response!=null)
```
Always use curly braces, even for one-line statements.

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoConnection.java:
```diff
@@ -26,14 +29,20 @@
         this.ip = ip;
         this.port = port;
-        connection = new Eiscp(ip, port);
+        connection = (EiscpInterface)new Eiscp(ip, port);
```
casting isn't necessary here, is it?

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoConnection.java:
```diff
@@ -26,14 +29,20 @@
 
+    public OnkyoConnection(String serialPortName) {
+        this.serialPortName = serialPortName;
+        connection = (EiscpInterface)new EiscpSerial(serialPortName);
```
casting isn't necessary here, is it?

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoBinding.java:
```diff
@@ -478,18 +485,28 @@ public String getHost() {
         OnkyoConnection getConnection() {
             if (connection == null) {
-                connection = new OnkyoConnection(host, port);
+                if (serialPortName!=null)
```
always use curly braces, even for one-line statements.

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoBinding.java:
```diff
@@ -478,18 +485,28 @@ public String getHost() {
+                if (serialPortName!=null)
+                    connection = new OnkyoConnection(serialPortName);
+                else
+                    connection = new OnkyoConnection(host, port);
```
always use curly braces, even for one-line statements.

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoBinding.java:
```diff
@@ -478,18 +485,28 @@ public String getHost() {
         @Override
         public String toString() {
-            return "Device [id=" + deviceId + ", host=" + host + ", port=" + port + "]";
+            if (serialPortName!=null)
```
always use curly braces, even for one-line statements.

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoBinding.java:
```diff
@@ -478,18 +485,28 @@ public String getHost() {
+            if (serialPortName!=null)
+                return "Device [id=" + deviceId + ", serialPort=" + serialPortName + "]";
+            else
+                return "Device [id=" + deviceId + ", host=" + host + ", port=" + port + "]";
```
always use curly braces, even for one-line statements.

watou commented:
Thank you very much, @basriram, for this enhancement.  I have review comments inline.  Some of these comments are automatically resolved when the source files are edited using the [new IDE setup](https://github.com/openhab/openhab-distro/blob/master/docs/sources/development/ide.md).  Please address the comments and squash your commits into 1 commit.  Thank you!

basriram commented:
@watou doing this for the first time so pardon the issues. I have fixed the items mentioned and squashed the two commits. Please let me know if I need to make further changes. Thx

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    private StringBuilder getEiscpMessage(String eiscpCmd) {
+        logger.debug("Requested Command is " + eiscpCmd);
```
Definitely use parameterized logging here instead of string concatenation, because otherwise this method will be generating garbage objects on every call!

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,258 @@
+        sb.append("1");
+        sb.append(eiscpCmd);
+        sb.append((char) 0x0D);
+        logger.debug("Serial command is " + sb);
```
Definitely use parameterized logging here instead of string concatenation, because otherwise this method will be generating garbage objects on every call!

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,258 @@
+    @Override
+    public void serialEvent(SerialPortEvent portEvent) {
+        try {
+            logger.debug("Received serial port event " + portEvent.toString() + ":" + portEvent.getEventType());
```
Definitely use parameterized logging here instead of string concatenation, because otherwise this method will be generating garbage objects on every call!

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    private String sendMmsg(String data, int timeout) throws IOException {
```
Is `sendMmsg` the proper name for this method?

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/eiscp/EiscpSerial.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    private String sendMmsg(String data, int timeout) throws IOException {
+        logger.debug("Sent request... " + data);
```
Definitely use parameterized logging here instead of string concatenation, because otherwise this method will be generating garbage objects on every call!

watou commented on bundles/binding/org.openhab.binding.onkyo/src/main/java/org/openhab/binding/onkyo/internal/OnkyoBinding.java:
```diff
@@ -253,8 +252,8 @@ public void updated(Dictionary<String, ?> config) throws ConfigurationException
                 if (!matcher.matches()) {
-                    logger.debug(
-                            "given config key '" + key + "' does not follow the expected pattern '<id>.<host|port>'");
+                    logger.debug("given config key '" + key
```
Definitely use parameterized logging here instead of string concatenation, because otherwise this method will be generating garbage objects on every call!

watou commented:
Hi @basriram, looks good, but with a few more comments (above).  

Could you also update [openhab_default.cfg](https://github.com/openhab/openhab/blob/master/distribution/openhabhome/configurations/openhab_default.cfg#L1057), [onkyo.cfg](https://github.com/openhab/openhab/blob/master/features/openhab-addons-external/src/main/resources/conf/onkyo.cfg), and add `<feature>openhab-transport-serial</feature>` as a dependent feature [here](https://github.com/openhab/openhab/blob/master/features/openhab-addons/src/main/feature/feature.xml#L361)?

After squashing your commits into one, I think this nice enhancement ought to be ready to merge.

Also, please update [the wiki page](https://github.com/openhab/openhab/wiki/Onkyo-Binding) with the new feature (noting that the serial support is only available starting in 1.9)?

basriram commented:
I have made all changes based on feedback and updated the Wiki. Please let me know if something else needs to be done before merging. Thx

watou commented:
Looks good, @basriram.  Could you squash your commits into 1 commit?

teichsta commented:
any updates @basriram? Would be great to have this PR merged.

basriram commented:
Hi,
Sorry I am not sure what I need to do as I am not real familiar with Git process. I thought I did manage to squash all my changes to a single commit before I did a pull request. If that was not correct,  please let me know what I need to do and I will try it again.
Thanks  

    On Sunday, March 6, 2016 1:52 PM, Thomas Eichstädt-Engelen <notifications@github.com> wrote:
 

 any updates @basriram? Would be great to have this PR merged.—
Reply to this email directly or view it on GitHub.

watou commented:
Hmm, your PR is from your master branch, so I'm not sure how to squash multiple commits into one.  But I'm sure there is a way using `git rebase`.  Right now this PR has 7 commits on it (see top of page or commits tab), but pull requests are easier to manage if they are made up of one commit.

basriram commented:
I had squashed all my changes into a single commit here - we can safely ignore the other 6 commits that happened prior to that. 

Merge branch 'master' of https://github.com/basriram/openhab · basriram/openhab@de2845e

|   |
|   |  |   |   |   |   |   |
| Merge branch 'master' of https://github.com/basriram/o...openhab - Source code of the open Home Automation Bus (openHAB) |
|  |
| View on github.com | Preview by Yahoo |
|  |
|   |


 

    On Sunday, March 6, 2016 4:36 PM, John Cocula <notifications@github.com> wrote:
 

 Hmm, your PR is from your master branch, so I'm not sure how to squash multiple commits into one. But I'm sure there is a way using git rebase. Right now this PR has 7 commits on it (see top of page or commits tab), but pull requests are easier to manage if they are made up of one commit.—
Reply to this email directly or view it on GitHub.

basriram commented:
I have created a new pull request with just a single commit #4202. Closing this one.
